src/dataflow-types/src/client.rs

```diff
@@ -1143,6 +1360,162 @@ pub mod tcp {
+        // is None, the consumer will wait on the queue_change notification to proceed only when a
+        // client has connected.
+        queue: Mutex<
+            Option<(
+                Receiver<super::ProtoComputeCommand>,
+                Sender<super::ProtoComputeResponse>,
+            )>,
+        >,
+
+        // If queue changes, a notification will be published here.
+        queue_change: Notify,
+    }
+
+    /// Server side implementation.
+    struct GrpcComputedServer {
+        // Keep going after first client has terminated connection
+        linger: bool,
+
+        // Send a signal here to shutdown this computed.
+        shutdown_tx: Sender<()>,
+
+        shared: Arc<GrpcComputedShared>,
+    }
+
+    /// Consumer side functions such as send and recv. These will not directly interact with the
+    /// network, but put messages in a mpsc queue which will be read and sent to the network in a
+    /// separate server task.
+    pub struct GrpcComputedServerInterface {
+        shared: Arc<GrpcComputedShared>,
+        // Communicate to the outer loop that the whole process should shut down. This channel will
+        // receive an empty message if and only if linger is not set and a client has disconnected.
+        //pub shutdown_rx: Receiver<()>,
+    }
+
+    impl GrpcComputedServerInterface {
+        pub async fn send(&self, resp: ComputeResponse) -> () {
+            loop {
+                match self.shared.queue.lock().await.as_ref() {
+                    Some(x) => {
+                        x.1.send((&resp).into_proto()).await.unwrap();
+                        break;
+                    }
+
+                    // Other end absent, wait for connection
+                    None => self.shared.queue_change.notified().await,
+                }
+            }
+        }
+
+        pub async fn recv(&mut self) -> Result<ComputeCommand, TryFromProtoError> {
+            loop {
+                match self.shared.queue.lock().await.as_mut() {
+                    Some(x) => match x.0.recv().await {
+                        Some(x) => return x.into_rust(),
+                        None => {
+                            println!("Recv returned None, endpoint gone, retry");
+                        }
+                    },
+
+                    // Other end absent, wait for connection
+                    None => {
+                        info!("recv called while no coordinator connected, waiting for coordinator connection.");
+                    }
+                }
+
+                self.shared.queue_change.notified().await;
+            }
+        }
+    }
+
+    type ResponseStream =
+        Pin<Box<dyn Stream<Item = Result<super::ProtoComputeResponse, tonic::Status>> + Send>>;
+
+    #[tonic::async_trait]
+    impl super::proto_compute_server::ProtoCompute for GrpcComputedServer {
+        type HandleStream = ResponseStream;
+
+        async fn handle(
```

Comment:
Nit: `handle` seems to be a very generic term. Could we find something more descriptive?

Reply:
hm i know. what about `setup_command_response_stream` ?

Reply:
Command_response_stream?

Reply:
I did this, following the google style guides (CamelCase for RPC services and functions).